### PR TITLE
Fix elixir-dogma on non-file-backed buffers

### PR DESCRIPTION
**Pull request by aaronjensen**
Split from #1063

### CODE REVIEW

File: flycheck.el [hunk 1]
```diff
@@ -6385,7 +6385,9 @@ This will either be the directory that contains `mix.exs' or,
 if no such file is found in the directory hierarchy, the directory
 of the current file."
   (or
-   (locate-dominating-file (buffer-file-name) "mix.exs")
+   (and
+    buffer-file-name
+    (locate-dominating-file buffer-file-name "mix.exs"))
```
Comment (kthelgason) [inline flycheck.el]: I'll repeat my comment from the other PR. Am I wrong in thinking that `buffer-file-name` is a function and it's result should be passed to `locate-dominating-file`?

Comment (cpitclaudel) [inline flycheck.el]: Partly wrong :) I replied there already.

Comment (cpitclaudel) [inline flycheck.el]:
```
buffer-file-name is a variable defined in ‘buffer.c’.

Documentation:
Name of file visited in current buffer, or nil if not visiting a file.
This should be an absolute file name.
```

**Comment by cpitclaudel** [comment]
Thanks! @kthelgason, what do you think?

File: flycheck.el (hunk 1 above)
Comment (aaronjensen) [inline flycheck.el]: `buffer-file-name` as a function is useful if you need the file name of a buffer other than the current. For the current you can just use the variable.

Comment (kthelgason) [inline flycheck.el]: Oh! Thanks for clarifying :)

**Comment by kthelgason** [comment]
LGTM :+1: Thanks @aaronjensen

File: flycheck.el (hunk 1 above)
Comment (lunaryorn) [inline flycheck.el]: I'm always using the function out of some habit that I can't explain :relaxed: but it's generally totally fine to use the variable instead.

**Comment by lunaryorn** [comment]
LGTM. @aaronjensen Feel free to merge :)

**Comment by cpitclaudel** [comment]
Sorry for the delay! LGTM, go ahead :)

**Pull request merged**